### ISSUE TITLE
qt: qmake problems due to BOM

### ISSUE THREAD

**Issue by fedepell** [ContinuumIO/anaconda-issues#4314]
Hi,
After upgrading the qt package from 5.6.2-3 to 5.6.2-5 I am having problems with variables reported by qmake for installation directory (which then make any build fail), namely:

`[root@abc ~]# /opt/anaconda3/bin/qmake -query QT_INSTALL_PREFIX`
`/opt/anaconda3/bin`

There is a /bin too much (that will spoil then all the other variables), while with previous 5.6.2-3 it was correct:

`[root@abc ~]# /opt/anaconda3/bin/qmake -query QT_INSTALL_PREFIX`
`/opt/anaconda3`

After some examination I found out that what it disturbs is the BOM marker (EF BB BF) that is now (since -4) at the very beginning of bin/qt.conf that wasn't present earlier. Removing that solves the issue.

If it may depend on the system: Anaconda Python 3.5 64-bit version running on a standard CentOS 7.2 distribution.

While the fix in my case may be trivial, is the presence of that BOM bytes important and/or why qmake doesn't handle it?

Thanks,
F.

**Comment by mingwandroid** [comment]
The BOM is important if you install into a prefix containing anything beyond ASCII. This seems to be a bug in qmake.exe since everything else in Qt handles the BOM just fine.

You can *probably* just delete `qt.conf`. It is not clear that that file is essential. Let me know if this also works for you.

**Comment by fedepell** [comment]
Yes, also if the file is deleted it will work correctly (at least as far as the example query done before is concerned).

**Comment by mingwandroid** [comment]
OK, does it also work if you create a new environment and delete the qt.conf from in there then try to run that qmake?

**Comment by fedepell** [comment]
No. In that case it will point to the system one, not anymore to the local environment as it should.

(so just to make sure I'm doing what you wanted me to try: I created a new environment as a normal user which reflects in a user .conda/envs/xxx directory, installed qt there and was placed correctly, qmake from the env was giving me the wrong path because of BOM, when I removed it then it will now point to /opt/anaconda3 and not to the local environment anymore as I suppose it should)

**Comment by mingwandroid** [comment]
Hmm, did you activate your new env too?

**Comment by fedepell** [comment]
Sorry I have to correct myself: I indeed activated the environment but there was some strange NFS interaction it would seem that was using the system one.

So: also deleting qt.conf from the newly created environment (or removing the BOM by hand) will work correctly and the variables from qmake -query will be correct.


One question: what is in general using qt.conf? I'm not so happy about removing it totally but it is also annoying to have to keep the BOM-less file managed myself on top.

Many Thanks!

**Comment by mingwandroid** [comment]
qt.conf is one way that Qt relocates itself and parts of itself. Thing it, I *believe* the values we emit are the default values anyway so it may serve no useful purpose.

**Comment by fedepell** [comment]
I have been browsing through qmake source (not in extreme detail, so please correct me if I'm wrong, I checked 5.7 and 5.9) and as far as I see (in property.cpp) for the values  QString::fromLatin1 / QString::toLatin1 are always used for setting and printing then the values. 
It looks to me that using UTF-8 in qt.conf would anyway be incorrectly handled from qmake in read/write (apart from the BOM itself) due to this.

Also as mentioned before my feeling is that the "standard" qt.conf I checked (3 different Linux versions and 2 Windows) comes always without the BOM in qt.conf. Is it something you added specifically to solve some issue with Anaconda maybe?

Many thanks again!

**Comment by mingwandroid** [comment]
If I can, I will drop qt.conf entirely. Otherwise I will patch qmake to skip (or better, handle correctly) the BOM.

**Comment by mingwandroid** [comment]
It isn't an issue with anaconda as such, more that it is impossible to relocate Qt, via qt.conf into a path that uses Unicode and we did not want this restriction to apply to the Anaconda Distribution (where people often install it in their home directory and that directory is often named as they are named, accented characters respected).

**Comment by fedepell** [comment]
Thanks again for digging into this!

I understand why you use BOM/Unicode for Anaconda and that's fair enough! I also agree that the problem is in qmake and should be solved there (upstream I'd say, not by the Anaconda as it's quiet an important change) by not assuming the Latin encoding but UTF8. I don't know if you have connections with the Qt company or in case I can give a look to report the issue and see if there is interest to fix (I find it strange it's there from long long time... I should try to do an installation from scratch of Qt in such a directory).

This being said: I found a workaround which may, if you think so, be an acceptable (dirty) solution for now: if you keep the BOM as you have now but put a newline (\n) just after the BOM (resulting in practise in an empty line at the head of the file) then qmake will accept it and read it fine. This is for how the file is parsed (line by line) and if you put this white line then the BOM will be eaten as part of this line and then the next (with the [Paths] section) will be parsed correctly and everything will work. Otherwise the first read line is [BOM][Paths] which doesn't match a valid section and everything is discarded.
Other applications using qt.conf should not be affected either (also as far as I checked), as it is just an empty line as in any INI-like file.

I know it's not the best solution but until qmake is fixed and we want to keep BOM in the file I think it's the best one we can get.

Cheers,
F.
